src/hooks/useGatewayAuthToken.ts

```diff
@@ -0,0 +1,73 @@
+import { Auth } from '@supabase/ui';
+import { useQuery, useSelect } from 'hooks/supabase-swr';
+import { client } from 'services/client';
+import { TABLES } from 'services/supabase';
+import useSWR from 'swr';
+import { GatewayAuthTokenResponse } from 'types';
+import {
+    getGatewayAuthTokenSettings,
+    getStoredGatewayAuthConfig,
+    getSupabaseAnonymousKey,
+    storeGatewayAuthConfig,
+} from 'utils/env-utils';
+
+interface CombinedGrantsExtQuery {
+    id: string;
+    object_role: string;
+}
+
+const { gatewayAuthTokenEndpoint } = getGatewayAuthTokenSettings();
+
+// The request body for this API is a string array corresponding to the prefixes a user has access to.
+const fetcher = (
+    endpoint: string,
+    prefixes: string[],
+    sessionKey: string | undefined
+): Promise<GatewayAuthTokenResponse[]> => {
+    const headers: HeadersInit = {};
+
+    const { supabaseAnonymousKey } = getSupabaseAnonymousKey();
+
+    headers.apikey = supabaseAnonymousKey;
+    headers.Authorization = `Bearer ${sessionKey}`;
+    headers['Content-Type'] = 'application/json';
+
+    return client(endpoint, {
+        data: { prefixes },
+        headers,
+    });
+};
+
+const useGatewayAuthToken = () => {
+    const { session } = Auth.useUser();
+
+    const combinedGrantsQuery = useQuery<CombinedGrantsExtQuery>(
+        TABLES.COMBINED_GRANTS_EXT,
+        { columns: `id, object_role` },
+        []
+    );
+    const { data: grants } = useSelect(combinedGrantsQuery);
+
+    const prefixes: string[] =
+        grants?.data.map(({ object_role }) => object_role) ?? [];
+
+    const gatewayConfig = getStoredGatewayAuthConfig();
+
+    return useSWR(
+        !gatewayConfig?.token && prefixes.length > 0
```

Comment:
This condition needs to be expanded. The `gateway_auth_token` API should be hit if there is no gateway configuration, no authentication token, or no valid authentication token in local storage and object role grants exist for this user. A JWT library will need to be used to determine whether an authentication token has expired. Any preferences?

Reply:
Not really - we should start by looking at Supabase and if they don't have one find one.